Properly isolate stacking contexts with mix-blend-mode children

rlhunt opened the pull request:
Anything in HTML that creates a stacking context must create an isolated group for mix-blend-mode. [1] This means that the initial backdrop for any children with mix-blend-mode must be transparent black. Currently stacking context's don't create isolated groups, and instead the backdrop can be from multiple parents up the stacking context tree.

This is a WIP that tries to implement the right behavior. When a stacking context has mix-blend-mode children, it's contents need to be drawn independently with the correct backdrop (transparent black), then composited.

It looks like the code works, but I'd like some feedback to see if this is an okay approach.

[1] https://www.w3.org/TR/compositing-1/#csscompositingrules_CSS

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/webrender/868)
<!-- Reviewable:end -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #867) made this pull request unmergeable. Please resolve the merge conflicts.

glennw commented:
@rlhunt I'm not quite sure I understand this - each stacking context in a mix-blend-mode case should already be isolated due to it being in a render target, shouldn't it?

I'm probably just missing something obvious here - but are you able to construct a HTML file that demonstrates this with a different rendering between Servo and Gecko perhaps?

rlhunt commented:
Yeah I can! Not sure the best way to link to a html file, so I'll just embed it. I adapted this from a gecko reftest.

``` html
<!DOCTYPE html>
<html>

<style type="text/css">
	.yellow {
		background-color: rgb(255, 255, 0);
	}
	.green {
		background-color: rgb(0, 255, 0);
	}
	.box {
		width: 100px;
		height: 100px;
	}
	.mbm {
		mix-blend-mode: difference;
	}
	.stacking-context {
		will-change: opacity;
	}
</style>

<!-- should see just the green box hiding the yellow box -->
<!-- the stacking context should create a new transparent backdrop for its children with mix-blend-mode -->
<!-- if this doesn't happen the green box will do difference with yellow creating red -->
<div class="yellow box">
	<div class="stacking-context">
		<div class="green box mbm">
		</div>
	</div>
</div>

</html>
```

emilio commented:
Note that this may also be a servo bug, since the `.stacking-context { will-change: opacity }` in Servo does _not_ force the creation of a stacking context. Display list dump:

```
┌ Display List
│  ├─ Items
│  │  ├─ PushStackingContext(StackingContext at TypedRect(1024px×740px at (0px,0px)) with overflow TypedRect(1024px×740px at (0px,0px)): StackingContextId(0)) StackingContext: StackingContextId(0) ScrollRoot: ScrollRootId(0)
│  │  ├─ SolidColor rgba(0, 0, 0, 0) @ TypedRect(1024px×740px at (0px,0px)) ClippingRegion::Max StackingContext: StackingContextId(0) ScrollRoot: ScrollRootId(0)
│  │  ├─ SolidColor rgba(0, 0, 0, 0) @ TypedRect(1008px×100px at (8px,8px)) ClippingRegion::Max StackingContext: StackingContextId(0) ScrollRoot: ScrollRootId(0)
│  │  ├─ SolidColor rgba(1, 1, 0, 1) @ TypedRect(100px×100px at (8px,8px)) ClippingRegion::Max StackingContext: StackingContextId(0) ScrollRoot: ScrollRootId(0)
│  │  ├─ SolidColor rgba(0, 0, 0, 0) @ TypedRect(100px×100px at (8px,8px)) ClippingRegion::Max StackingContext: StackingContextId(0) ScrollRoot: ScrollRootId(0)
│  │  ├─ PushStackingContext(StackingContext at TypedRect(100px×100px at (8px,8px)) with overflow TypedRect(100px×100px at (0px,0px)): StackingContextId(140394916598976)) StackingContext: StackingContextId(140394916598976) ScrollRoot: ScrollRootId(0)
│  │  ├─ SolidColor rgba(0, 1, 0, 1) @ TypedRect(100px×100px at (0px,0px)) ClippingRegion::Max StackingContext: StackingContextId(140394916598976) ScrollRoot: ScrollRootId(0)
│  │  ├─ PopStackingContext(StackingContextId(140394916598976) StackingContext: StackingContextId(140394916598976) ScrollRoot: ScrollRootId(0)
│  │  └─ PopStackingContext(StackingContextId(0) StackingContext: StackingContextId(0) ScrollRoot: ScrollRootId(0)
```

rlhunt commented:
Okay, I'm able to observe the same differences between servo and gecko by changing to `.stacking-context { position: absolute; z-order: 1 }` 

```
┌ Display List
│  ├─ Items
│  │  ├─ PushStackingContext(StackingContext at TypedRect(1024px×740px at (0px,0px)) with overflow TypedRect(1024px×740px at (0px,0px)): StackingContextId(0)) StackingContext: StackingContextId(0) ScrollRoot: ScrollRootId(0)
│  │  ├─ SolidColor rgba(0, 0, 0, 0) @ TypedRect(1024px×740px at (0px,0px)) ClippingRegion::Max StackingContext: StackingContextId(0) ScrollRoot: ScrollRootId(0)
│  │  ├─ SolidColor rgba(0, 0, 0, 0) @ TypedRect(1008px×100px at (8px,8px)) ClippingRegion::Max StackingContext: StackingContextId(0) ScrollRoot: ScrollRootId(0)
│  │  ├─ SolidColor rgba(1, 1, 0, 1) @ TypedRect(100px×100px at (8px,8px)) ClippingRegion::Max StackingContext: StackingContextId(0) ScrollRoot: ScrollRootId(0)
│  │  ├─ PushStackingContext(StackingContext at TypedRect(100px×100px at (8px,8px)) with overflow TypedRect(100px×100px at (0px,0px)): StackingContextId(140413524628528)) StackingContext: StackingContextId(140413524628528) ScrollRoot: ScrollRootId(0)
│  │  ├─ SolidColor rgba(0, 0, 0, 0) @ TypedRect(100px×100px at (0px,0px)) ClippingRegion::Max StackingContext: StackingContextId(140413524628528) ScrollRoot: ScrollRootId(0)
│  │  ├─ PushStackingContext(StackingContext at TypedRect(100px×100px at (0px,0px)) with overflow TypedRect(100px×100px at (0px,0px)): StackingContextId(140413524628672)) StackingContext: StackingContextId(140413524628672) ScrollRoot: ScrollRootId(0)
│  │  ├─ SolidColor rgba(0, 1, 0, 1) @ TypedRect(100px×100px at (0px,0px)) ClippingRegion::Max StackingContext: StackingContextId(140413524628672) ScrollRoot: ScrollRootId(0)
│  │  ├─ PopStackingContext(StackingContextId(140413524628672) StackingContext: StackingContextId(140413524628672) ScrollRoot: ScrollRootId(0)
│  │  ├─ PopStackingContext(StackingContextId(140413524628528) StackingContext: StackingContextId(140413524628528) ScrollRoot: ScrollRootId(0)
│  │  └─ PopStackingContext(StackingContextId(0) StackingContext: StackingContextId(0) ScrollRoot: ScrollRootId(0)
```

glennw commented:
Ah, I see what you mean now, thanks for the example! I'll take a look over the PR today.

glennw commented:
The general idea here seems quite good, I think. It seems a little unfortunate to have to modify frame.rs and the conditions used are also a bit confusing (I guess the next_z == 0 is a way of checking if at the root?).

I wonder if we could remove the frame.rs changes, and handle this all internally within FrameBuilder? I'd be completely fine with FrameBuilder internally keeping a stack of the current stacking context indices. I think that would make it feasible to implement completely inside FrameBuilder?

It also might make sense to make the concept of whether a stacking context needs to be isolated a little bit more generic, rather than naming things specifically related to having mix blend mode children?

Sorry that's a bit vague - just some initial thoughts. Thanks for looking into this!

glennw commented:
@rlhunt Do the comments / questions above make sense? We could discuss on IRC / Vidyo if that would be easier?

rlhunt commented:
@glennw Yes they do make sense! I agree that the conditions could be more clear and cleaned up. I like the idea of moving this to FrameBuilder. It can probably be done completely in there.

I've been side tracked with other tasks, but I'll try and take a look at this sometime soon!

glennw commented:
No worries, thanks for looking into this stuff! :)

rlhunt commented:
@glennw Okay, I've rebased this and cleaned it up. Now all the logic for when to isolate a stacking context happens in frame_builder.rs, and I've tried to add a few comments to explain what 'should_isolate' means and the conditions for it.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #904) made this pull request unmergeable. Please resolve the merge conflicts.

glennw commented on webrender/src/frame_builder.rs:
```diff
@@ -157,6 +159,17 @@ impl FrameBuilder {
+        self.stacking_context_stack.push(stacking_context_index);
+
+        if let Some(parent_index) = parent_index {
+          if composite_ops.mix_blend_mode.is_some() {
```
nit: should be 4 space indent

glennw commented:
@rlhunt OK, seeing some test failures that appear related:

```
  ▶ FAIL [expected PASS] /compositing-1_dev/html/compositing_simple_div.htm
  └   → /compositing-1_dev/html/compositing_simple_div.htm 13ee95e760e850374940f5c9633148e0574e0287
/compositing-1_dev/html/reference/compositing_simple_div-ref.htm 66bf04c1597f2bdb7a831172129d16d8e21bea17
Testing 13ee95e760e850374940f5c9633148e0574e0287 == 66bf04c1597f2bdb7a831172129d16d8e21bea17

  ▶ FAIL [expected PASS] /compositing-1_dev/html/mix-blend-mode-stacking-context-creates-isolation.htm
  └   → /compositing-1_dev/html/mix-blend-mode-stacking-context-creates-isolation.htm 33fee4db992ec9f203e9c4b45a7e83ebea8fbcfe
/compositing-1_dev/html/reference/mix-blend-mode-stacking-context-creates-isolation-ref.htm e712a73bb530ce05bad070100c1831f6795f6bcd
Testing 33fee4db992ec9f203e9c4b45a7e83ebea8fbcfe == e712a73bb530ce05bad070100c1831f6795f6bcd
```

You can investigate these locally by building Servo with a local WR override, and then running:

```
./mach test-css --release --processes=6
```

glennw commented:
(I should note - it's possible that these were accidental passes previously - I haven't compared the reftest output before / after this PR on those tests).

rlhunt commented:
Thanks for running the tests! I'll take a look at those failures.

rlhunt commented:
Okay, so I think I've figured out the test failures. I'll update with a patch sometime soon.

rlhunt commented:
So by soon I guess I meant almost a week :)

There are two major problems
1. We aren't taking the backdrop's alpha into account ([spec](https://drafts.fxtf.org/compositing-1/#generalformula))
2. The backdrop readback is upside down when we aren't doing readback from the framebuffer

I've updated and rebased to take both into account. I've removed the hardware blend mode path because I couldn't think of how to incorporate the backdrop's alpha. I ran it locally on a build of servo and it didn't seem to have any related failures.

rlhunt commented:
I should also note that '/compositing-1_dev/html/compositing_simple_div.htm' displays differently in Chrome vs Firefox. Chrome has the div blend with the canvas background while Firefox does not.

Chrome seems to have the expected behavior for the test so that is the what I went with.

glennw commented:
@bors-servo r+

Thanks for following up on this!

bors-servo commented:
:clipboard: Looks like this PR is still in progress, ignoring approval

Manishearth commented:
@bors-servo r=gw

bors-servo commented:
:pushpin: Commit 5100490 has been approved by `gw`

<!-- @bors-servo r=gw 51004908d676dbd956bc6f6b715b39c8602f12ba -->

bors-servo commented:
:hourglass: Testing commit 5100490 with merge 040805f...

bors-servo commented:
:sunny: Test successful - [status-travis](https://travis-ci.org/servo/webrender/builds/206453627)
Approved by: gw
Pushing 040805f31650527311c0a0d58ccf9620b0f3eca9 to master...
